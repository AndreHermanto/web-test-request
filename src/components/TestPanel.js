import React, { Component } from 'react';
import {
  FormGroup,
  ControlLabel,
  Row,
  Col,
  Label
} from 'react-bootstrap';
import { 
  SubLabel,
  ValidationFeedback
} from './SharedStyle';
import styled from 'styled-components';
import checked from '../assets/images/checked.svg';
import grayCheck from '../assets/images/checked-gray.svg';

const Gene = styled(Label)`
  margin-right: 4px;
  display: inline-block !important;
  font-weight: 300 !important;
`;

const TestContainer = styled.div`
  padding: 12px !important;
  border-radius: 0px;
  background-color: #fff;
  border: 1px solid #ccc;
  margin-top: 10px;
  margin-bottom: 10px;
  cursor: pointer;
  color:${props => props.valid !== 'error' ? 'black' : '#A94445'};
  &:hover {
    border-color:#00a6b6;
    color: #00a6b6 !important;
  }

  &.active {
    border-color:#00a6b6;
    color: #00a6b6 !important;
  }
`;

/**
 * This creates a set of gene label buttons via a single component.
 */
class TestPanel extends Component {
  constructor() {
    super();
    this.state = {
      image:''
    }
    this.getValidator = this.getValidator.bind(this);
    this.getLabel = this.getLabel.bind(this);
    this.getGene = this.getGene.bind(this);
    this.handleOnClick = this.handleOnClick.bind(this);
  }

  getValidator() {
    if(!this.props.onValidate) {
      return { status: null, rule: {} };
    } else {
      return this.props.onValidate[this.props.field] || { status: null, rule: {} };
    }
  }

  getLabel() {
    return <ControlLabel>
      {this.props.label}
      <SubLabel>Required</SubLabel>
    </ControlLabel>
  }

  getGene(genes) {
    let gene = [];
    if(genes !== undefined) {
      genes.map((g, i) =>{
        if(i < 3) return gene.push(g);
        return gene;
      })
      return gene;
    }
  }

  handleOnClick(panel) {
    this.props.handleClick(panel);
  }

  render() { 
    let validator = this.getValidator();
    let label = this.getLabel();
    return (
      <FormGroup  validationState={validator.status}>
        {label}
        <br/>
        {
          this.props.options.length > 0 && 
          this.props.options.map((option, i) =>
          {
            let genes = this.getGene(option.genes);
            let totalGene;
            let panels;
            if(option.genes !== undefined)
            {
              totalGene = option.genes.length - 3;
            }
            if(this.props.latestSelectId === option.id)
            {
              panels = <Panel key={i} image={checked} geneList={this.props.genes}
              totalGene={totalGene} option={option} genes={genes} valid={validator.status} handleClick={this.handleOnClick}/>;;
            }
            else {
              panels = <Panel key={i} image={grayCheck} geneList={null} 
              totalGene={totalGene} option={option} genes={genes} valid={validator.status} handleClick={this.handleOnClick}/>;
            }
            return panels;
          })
        }
        <ValidationFeedback>{validator.feedback}</ValidationFeedback> 
      </FormGroup>
    );
  }
}

function Panel(props) {
  let geneList;
  if(props.geneList !== undefined && props.geneList !== null) {
    geneList = props.geneList.map((gene, $index) => {
      return <Gene key={$index}>{gene}</Gene> 
    })
  }
  else {
    geneList = '';
  }
  return (
    <TestContainer
      valid={props.valid}
      onClick={() => props.handleClick(props.option)}>
      <Row style={{ padding: '9px' }}>
        <Col md={1}>
          <img 
            src={props.image} 
            alt="logo" 
            style={{ height:40, marginLeft: -4 }} />
        </Col>
        <Col md={11}>
          {
            props.option.label === 'Whole Genome Analysis' ?
            <div style={{marginTop:10}}>{props.option.label}</div> :
            <div>{props.option.label}</div>
          }
          {
            props.genes !== undefined &&
            <div className="text-muted" style={{ fontSize: 12.5 }}>
              { 
                props.genes.map((g, i) => {
                  return (<span key={i}> {g}</span>)
                })
              }
              {
                props.totalGene > 0 &&
                <span> + {props.totalGene} more </span>
              }
            </div>
          }
        </Col>
      </Row>
      <Row>
        <Col md={12}>
          { 
            (props.geneList !== null && props.geneList !== undefined) &&
            <div style={{ padding: '8px 8px 0 8px' }}> 
              <ControlLabel>Available Genes:</ControlLabel>
              <div>{geneList}</div>
            </div>
          }
        </Col>
      </Row>
    </TestContainer>
  )
}

export default TestPanel;