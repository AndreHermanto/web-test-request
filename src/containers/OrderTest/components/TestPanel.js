import React, { Component } from 'react';
import {
  FormGroup,
  ControlLabel,
  Label,
  Row,
  Col,
  Radio
} from 'react-bootstrap';
import { 
  SubLabel,
  ValidationFeedback
} from './../../../components/SharedStyle';
import styled from 'styled-components';
import checked from './../../../assets/images/checked.svg';
import grayCheck from './../../../assets/images/checked-gray.svg';

const Gene = styled(Label)`
  margin-right: 4px;
  display: inline-block !important;
  font-weight: 300 !important;
  background: ${props => props['data-inCore'] ? '#7788aa' : '#00a6b6'} !important;
`;

const TestContainer = styled.div`
  padding: 11px 0px 11px 21px !important;
  border-radius: 0px;
  background-color: #fff;
  border: 1px solid #ccc;
  margin-top: 6px;
  margin-bottom: 6px;
  cursor: pointer;
  line-height: 1.2;
  color:${props => props.selected ? '#00a6b6' : '#000'};
  border-color: ${props => props.selected ? '#00a6b6' : '#ccc'};
  &:hover {
    border-color: #00a6b6;
    color: #00a6b6 !important;
  }

  &.active {
    border-color: #00a6b6;
    color: #00a6b6 !important;
  }

  -webkit-box-shadow: inset 7px 0px 0px 0px ${props => props.selected ? '#00a6b6' : '#ccc'};
  -moz-box-shadow: inset 7px 0px 0px 0px ${props => props.selected ? '#00a6b6' : '#ccc'};
  box-shadow: inset 7px 0px 0px 0px ${props => props.selected ? '#00a6b6' : '#ccc'};
`;

const Tag = styled(Label)`
  float: right;
  margin-right: 12px;
  padding: 4px 8px 3px 8px !important;
  background: ${props => props.type === 'core' ? '#00d6e6' : '#007580'} !important;
`; 

/**
 * This creates a set of gene label buttons via a single component.
 */
class TestPanel extends Component {
  constructor(props) {
    super(props);

    this.getValidator = this.getValidator.bind(this);
    this.getLabel = this.getLabel.bind(this);
    this.handleOnClick = this.handleOnClick.bind(this);
  }

  getValidator() {
    if(!this.props.onValidate) {
      return { status: null, rule: {} };
    } else {
      return this.props.onValidate[this.props.field] || { status: null, rule: {} };
    }
  }

  getLabel() {
    return <ControlLabel>
      {this.props.label}
      <SubLabel>Required</SubLabel>
    </ControlLabel>
  }

  handleOnClick(panel) {
    this.props.handleClick(panel);
  }

  render() { 
    let validator = this.getValidator();
    let label = this.getLabel();
    return (
      <FormGroup validationState={validator.status}>
        {label}
        <br/>
        {
          (this.props.options && this.props.options.length > 0) && 
          this.props.options.map((option, i) => {
            let panels;

            if(this.props.preSelect.id === option.id) {
              panels = <Panel
                        key={i}
                        image={checked} 
                        option={option}
                        valid={validator.status}
                        panelClick={this.handleOnClick}
                        typeSelect={this.props.handleTypeClick}
                        preSelectType={this.props.preSelect.geneLists[0] ? this.props.preSelect.geneLists[0].type : 'complete'}
                        formState={this.state}
                        selected={true}
                      />;
            } else {
              panels = <Panel
                        key={i}
                        image={grayCheck}
                        option={option}
                        valid={validator.status}
                        panelClick={this.handleOnClick}
                        typeSelect={this.props.handleTypeClick}
                        preSelectType={this.props.preSelect.geneLists[0] ? this.props.preSelect.geneLists[0].type : 'complete'}
                        formState={this.state}
                        selected={false}
                      />;
            }
      
            return panels;
          })
        }
        <br/>
        <ValidationFeedback>{validator.feedback}</ValidationFeedback> 
      </FormGroup>
    );
  }
}

function determineTag(label) {
  if(label.search('Core Panel') !== -1) return <Tag type="core">Core</Tag>;
  if(label.search('Complete Panel') !== -1) return <Tag type="complete">Complete</Tag>;
  return '';
}

function Panel(props) {
  // Process in a way that the format would work with RadioSet module - needs id and label.
  var geneList = props.option.geneLists ? props.option.geneLists.slice(0) : [];
  geneList.map((list) => {
    list.id = list.type;
    list.label = list.type.toUpperCase().slice(0,1) + list.type.slice(1);   
    return list;
  })

  return (
    <TestContainer
      valid={props.valid}
      selected={props.selected}
      label={props.option.label}
      onClick={() => props.panelClick(props.option)}
    >
      <div>
        <img 
          src={props.image} 
          alt="logo" 
          style={{ height:30, float: 'left'}} />

        {determineTag(props.option.label)}

        <div style={{ 
          height: 30, 
          verticalAlign: 'middle', 
          display: props.selected ? 'initial' : 'table-cell', 
          paddingLeft: 20, 
          paddingRight: 20 
        }}>
        
          {props.option.label}
        </div>
      </div>
        
      {
        props.selected &&
        <div style={{ marginTop: 32 }}>
          {
            geneList.map((list, $index) => {
              return <Row key={$index} style={{ padding: 8 }}>
                <Col lg={3}>
                  <Radio
                    style={{ margin: 0, color: list.type === 'core' ? '#7788aa' : '#00a6b6'}}
                    key={$index}
                    name="type" 
                    value={list.type}
                    onClick={(event) => props.typeSelect(event, list)}
                    defaultChecked={list.type === props.preSelectType}
                  >
                    {list.label} Panel <small style={{ display: 'inline-block' }}>({list.genes.length} genes included)</small>
                  </Radio>
                </Col>
                <Col lg={9}>
                  {list.genes.map((gene, $geneIndex) => {
                    // detect whether the gene is in both core and complete.
                    var inCore = false; 
                    if(geneList.length > 1) {
                      geneList[0].genes.forEach((coregene) => {
                        if(coregene === gene) inCore = true;
                      });
                    }
                    return <Gene key={$geneIndex} data-inCore={inCore} >{gene}</Gene> 
                  })}
                </Col>
              </Row>
            })
          }
        </div>
      }
    </TestContainer>
  )
}

export default TestPanel;