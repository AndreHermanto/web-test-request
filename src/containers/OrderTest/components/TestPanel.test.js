import React from 'react';
import TestPanel from './TestPanel';
import renderer from 'react-test-renderer';
import TestUtils from 'react-addons-test-utils';

var props = {
  handleClick: jest.fn(),
  options: [
    { id: 'p1', label: 'p1', genes: ['AAA','BBB','CCC'] },
    { id: 'p2', label: 'p2', genes: ['DDD','EEE'] },
    { id: 'p3', label: 'p3', genes: ['FFF'] }
  ]
}

describe('TestPanel', function() {
  test('TestPanel overall renders well', async function() {
    const page = renderer.create(React.createElement(TestPanel, props)).toJSON();
    expect(page).toMatchSnapshot();
  });
});