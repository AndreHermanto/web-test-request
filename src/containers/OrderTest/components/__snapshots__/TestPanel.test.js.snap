exports[`TestPanel TestPanel overall renders well 1`] = `
<div
  className="form-group">
  <label
    className="control-label"
    htmlFor={undefined}>
    <span
      className="lnMAUb">
      Required
    </span>
  </label>
  <br />
  <div
    className="iRRKpJ"
    label="p1"
    onClick={[Function]}
    selected={false}>
    <div>
      <img
        alt="logo"
        src="test-file-stub"
        style={
          Object {
            "float": "left",
            "height": 30,
          }
        } />
      
      <div
        style={
          Object {
            "display": "table-cell",
            "height": 30,
            "paddingLeft": 20,
            "paddingRight": 20,
            "verticalAlign": "middle",
          }
        }>
        p1
      </div>
    </div>
  </div>
  <div
    className="iujgGu"
    label="p2"
    onClick={[Function]}
    selected={true}>
    <div>
      <img
        alt="logo"
        src="test-file-stub"
        style={
          Object {
            "float": "left",
            "height": 30,
          }
        } />
      
      <div
        style={
          Object {
            "display": "initial",
            "height": 30,
            "paddingLeft": 20,
            "paddingRight": 20,
            "verticalAlign": "middle",
          }
        }>
        p2
      </div>
    </div>
    <div
      style={
        Object {
          "marginTop": 32,
        }
      }>
      <div
        className="row"
        style={
          Object {
            "padding": 8,
          }
        }>
        <div
          className="col-lg-3">
          <div
            className="radio"
            style={
              Object {
                "color": "#00a6b6",
                "margin": 0,
              }
            }>
            <label>
              <input
                defaultChecked={false}
                disabled={false}
                name="type"
                onClick={[Function]}
                type="radio"
                value="complete" />
              Complete
               Panel 
              <small
                style={
                  Object {
                    "display": "inline-block",
                  }
                }>
                (
                3
                 genes included)
              </small>
            </label>
          </div>
        </div>
        <div
          className="col-lg-9">
          <span
            className="hynVdE label label-default"
            data-inCore={true}>
            DDD
          </span>
          <span
            className="hynVdE label label-default"
            data-inCore={true}>
            EEE
          </span>
          <span
            className="hynVdE label label-default"
            data-inCore={true}>
            FFF
          </span>
        </div>
      </div>
      <div
        className="row"
        style={
          Object {
            "padding": 8,
          }
        }>
        <div
          className="col-lg-3">
          <div
            className="radio"
            style={
              Object {
                "color": "#7788aa",
                "margin": 0,
              }
            }>
            <label>
              <input
                defaultChecked={true}
                disabled={false}
                name="type"
                onClick={[Function]}
                type="radio"
                value="core" />
              Core
               Panel 
              <small
                style={
                  Object {
                    "display": "inline-block",
                  }
                }>
                (
                1
                 genes included)
              </small>
            </label>
          </div>
        </div>
        <div
          className="col-lg-9">
          <span
            className="hynVdE label label-default"
            data-inCore={true}>
            FFF
          </span>
        </div>
      </div>
    </div>
  </div>
  <br />
  <span
    className="SfhHK help-block" />
</div>
`;
